### PR TITLE
[JENKINS-52888] Add option to use S3 transfer acceleration

### PR DESCRIPTION
**Pull request by alreay**
As per description; adds support for using s3 transfer acceleration . It can also create buckets with this feature enabled.

### CODE REVIEW

**Comment by jglick** [comment]
Is this similar to #75, addressing JENKINS-52888?

**Comment by alreay** [comment]
Hi Jesse, yes, same intent as #75 . Hopefully this PR resolves some of the problems that you had with that one

**Comment by alreay** [comment]
Hi @jglick , just wondering if there's anything that you're waiting on me for this? I see the review check failed for java11 but looks like that's failing for everything.

**Comment by jglick** [comment]
No, just waiting for me (or some new maintainer) to give this plugin some love.

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStore.java [hunk 1]
```diff
@@ -262,6 +262,7 @@ public String toString() {
         sb.append(", region='").append(getRegion()).append('\'');
         sb.append(", deleteArtifacts='").append(isDeleteArtifacts()).append('\'');
         sb.append(", deleteStashes='").append(isDeleteStashes()).append('\'');
+        sb.append(", useTransferAcceleration='").append(getConfiguration().getUseTransferAcceleration()).append('\'');
```
Comment (Vlatombe) [inline src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStore.java:265]: The new field is not even used in this class.
```suggestion
```

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java [hunk 2]
```diff
@@ -341,8 +359,14 @@ public Bucket createS3Bucket(String name) throws IOException {
 
     private Bucket createS3Bucket(String name, boolean disableSessionToken) throws IOException {
         AmazonS3ClientBuilder builder = getAmazonS3ClientBuilderWithCredentials(disableSessionToken);
-        AmazonS3 client = builder.build();
-        return client.createBucket(name);
+        //Creating a new bucket cannot use transfer acceleration
+        AmazonS3 client = builder.withAccelerateModeEnabled(false).build();
```
Comment (Vlatombe) [inline src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java:363]: AFAICS, this is not useful, the bucket can be created just fine keeping the previous implementation.

**Review by Vlatombe — CHANGES_REQUESTED** [review]
If transfer acceleration is enabled, the configuration validation should ensure it is enabled on the S3 bucket, otherwise builds will fail to archive with the following message `S3 Transfer Acceleration is not configured on this bucket`.

See also some remarks on the implementation. Otherwise, the feature works.

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStore.java (hunk 1 above)
Comment (alreay) [reply src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStore.java]: Removed

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java (hunk 2 above)
Comment (alreay) [reply src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStoreConfig.java]: Correct, the bucket can be created while using accelerated mode. Unfortunately, for some reason, you can't set the accelerated mode on a bucket while your client is in accelerated mode. I've updated the comment to make that clearer

**Comment by boussaffawalid** [comment]
I tested this PR on our our prod environment and it works as expected. Is there plans to merge this PR at some point?

File: src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStore.java (hunk 1 above)
Comment (alreay) [reply src/main/java/io/jenkins/plugins/artifact_manager_jclouds/s3/S3BlobStore.java]: If you use this feature with an S3 bucket fer acceleration the "Validate S3 Bucket Configuration" button on the plugin config page gives this error
> GetBucketLocation failed
> com.amazonaws.services.s3.model.AmazonS3Exception: S3 Transfer Acceleration is disabled on this bucket

**Review by jglick — APPROVED** [review]
Seems to break nothing if disabled. I enabled acceleration on a test bucket and confirmed that validation then succeeded and I was able to run a build archiving a tiny artifact. The resulting presigned URL mentioned acceleration so I suppose it worked.

**Comment by boussaffawalid** [comment]
@alreay @jglick Artifact Manager on S3 does not delete artifacts from s3 when use Use Transfer Acceleration is enabled. I opened a ticket here https://issues.jenkins.io/browse/JENKINS-70252

According to aws documentation:

> You can use all of the Amazon S3 operations through the transfer acceleration endpoints, except for the following operations: [GET Service (list buckets)](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTServiceGET.html), [PUT Bucket (create bucket)](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUT.html), and [DELETE Bucket](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketDELETE.html). Also, Amazon S3 Transfer Acceleration does not support cross region copies using [PUT Object - Copy](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html).
 
so so it seems like withAccelerateModeEnabled needs to be set to false before deleting blobs from s3.

**Pull request merged**